File: src/test/python/datastructures/linkedlist/linked_list_test.py
import unittest
from linked_list import LinkedList
import random

class TestLinkedList(unittest.TestCase):

    def setUp(self):
        self.lst = LinkedList()
        self.MAX_RAND_NUM = 250
        self.TEST_SZ = 50
        self.rand_elem = random.randrange(self.MAX_RAND_NUM)
        
    def test_size(self):
        # test empty linked list size method.
        self.assertEqual(self.lst.size(), 0)

        for i in range(self.TEST_SZ):
            self.lst.add(i)
        self.assertEqual(self.lst.size(), self.TEST_SZ)
    
    def test_is_empty(self):
        self.assertTrue(self.lst.is_empty())
        self.lst.add(random.randrange(self.MAX_RAND_NUM))
        self.assertFalse(self.lst.is_empty())

    def test_contains(self):
        # check false for empty list, true once element is added.
        self.assertFalse(self.lst.contains(self.rand_elem))
        self.lst.add(self.rand_elem)
        self.assertTrue(self.lst.contains(self.rand_elem))

    def test_peek(self):
        # assert None on empty list.
        self.assertIsNone(self.lst.peek())

        # assert correct item after adding multiple items.
        rand_list = random.sample(range(self.MAX_RAND_NUM), self.TEST_SZ)
        last_item_added = rand_list[len(rand_list) - 1]
        for elem in rand_list:
            self.lst.add(elem)
        self.assertEqual(self.lst.peek(), last_item_added)

    def test_add(self):
        # add on empty list:
        self.lst.add(self.rand_elem)
        self.assertEqual(self.lst.peek(), self.rand_elem)
        self.assertEqual(self.lst.size(), 1)

        # add many items on non-empty list:
        rand_list = random.sample(range(self.MAX_RAND_NUM), self.TEST_SZ)
        last_item_added = rand_list[len(rand_list) - 1]
        for elem in rand_list:
            self.lst.add(elem)
        self.assertEqual(self.lst.peek(), last_item_added)
        self.assertEqual(self.lst.size(), self.TEST_SZ + 1)

    def test_add_last(self):
        # add on empty list.
        self.lst.add_last(self.rand_elem)
        self.assertEqual(self.lst.peek(), self.rand_elem)

        # add many items non-empty list:
        rand_list = random.sample(range(self.MAX_RAND_NUM), self.TEST_SZ)
        for elem in rand_list:
            self.lst.add_last(elem)
        self.assertEqual(self.lst.size(), self.TEST_SZ + 1)
        
    def test_remove(self):
        self.lst.add_last(self.rand_elem)
        self.lst.remove(self.rand_elem)
        self.assertFalse(self.lst.contains(self.rand_elem))
    
    def test_pop(self):
        self.lst.add(self.rand_elem)
        self.assertEqual(self.lst.pop(), self.rand_elem)
        self.assertEqual(self.lst.size(), 0 )
         
if __name__ == '__main__':
    unittest.main()

File: src/test/ruby/datastructures/linkedlist/linked_list_test.rb
require "./src/main/ruby/datastructures/linkedlist/linked_list"
require "test/unit"

class TestLinkedList < Test::Unit::TestCase

    def setup
        @list = LinkedList.new
        @TEST_SZ = 50
    end

    def test_new_sllist_size
        # test empty sllist
        assert_equal(0, @list.size)

        # test smallest non-empty sllist
        @list.push(3)
        assert_equal(1, @list.size)

        # test a higher bound 
        for i in 1..@TEST_SZ
            @list.push(i)
        end
        assert_equal(51, @list.size)
    end

    def test_is_empty
        assert(@list.is_empty)

        @list.push(3)
        assert(!@list.is_empty)
    end

    def test_contains_one_item
        assert(!@list.contains(3))
        @list.push(3)
        assert(@list.contains(3))
    end

    def test_contains_multiple_items
        for i in 1..@TEST_SZ
            @list.push(i)
        end
        
        for i in 1..@TEST_SZ
            assert(@list.contains(i))

        end
    end
    def test_push_one_item
        @list.push(5)
        assert_equal(1, @list.size)
        assert(!@list.contains(2)) # check it does not contain a different number
        assert(@list.contains(5))
    end

    def test_push_multiple_items
        lst_one = [1,3,5,7,9]
        lst_two = [2,4,6,8,10]
        for num in lst_one
            @list.push(num)
        end
        assert_equal(5, @list.size)
        
        # check list contains items pushed
        for num in lst_one
            assert(@list.contains(num))
        end
        # check no incorrect numbers were added
        for num in lst_two
            assert(!@list.contains(num))
        end
    end

    def test_pop_one_item
        assert_raise(IndexError){
            @list.pop
        }
        @list.push(1)
        assert_equal(1, @list.pop)
        assert_equal(0, @list.size)
    end

    def test_pop_multiple_items
        for i in 1..@TEST_SZ
            @list.push(i)
        end
        assert_equal(@TEST_SZ, @list.pop)
    end
end